apps/hashdotdev/src/components/mdx-page-content.tsx

```diff
@@ -1,16 +1,132 @@
-// @todo update from blockprotocol
+// import throttle from "lodash/throttle";
+import { useRouter } from "next/router";
 import { MDXRemote, MDXRemoteSerializeResult } from "next-mdx-remote";
-import { FunctionComponent } from "react";
+import { FunctionComponent, useEffect, useMemo, useRef, useState } from "react";
 
 import { mdxComponents } from "../util/mdx-components";
+import { Heading, PageHeadingsContext } from "./context/page-headings-context";
 
 type MdxPageContentProps = {
   serializedPage: MDXRemoteSerializeResult<Record<string, unknown>>;
 };
 
+let detectHeadingFromScrollTimer: NodeJS.Timeout | undefined = undefined;
+
 export const MdxPageContent: FunctionComponent<MdxPageContentProps> = ({
   serializedPage,
 }) => {
-  // @ts-expect-error @todo fix this
-  return <MDXRemote {...serializedPage} components={mdxComponents} />;
+  const router = useRouter();
+
+  const [headings, setHeadings] = useState<Heading[]>([]);
+
+  const detectHeadingFromScroll = useRef<boolean>(true);
+
+  const currentHeading = useRef<Heading | undefined>(undefined);
+
+  const headingsRef = useRef<Heading[]>([]);
+
+  useEffect(() => {
+    setHeadings([]);
+
+    return () => {
+      currentHeading.current = undefined;
+      setHeadings([]);
+    };
+  }, [serializedPage]);
+
+  const scrolledOnce = useRef(false);
+
+  useEffect(() => {
+    if (headings.length) {
+      const anchor = router.asPath.match(/(?:#)(.*?)(?:\?|$)/)?.[1] ?? "";
+
+      const headingWithCurrentAnchor = headings.find(
+        (heading) => heading.anchor === anchor,
+      );
+
+      let previousRoute;
+      try {
+        previousRoute = window.sessionStorage.getItem("previousRoute");
+      } catch {
+        // sessionStorage is not available
+      }
+
+      const shouldScrollToAnchor =
+        // if anchor is empty and we haven't scrolled, prevent it
+        (anchor === "" && scrolledOnce.current) ||
+        // if anchor is not empty, always allow scroll
+        anchor !== "" ||
+        // OR if previous path is either a docs or spec page
+        (previousRoute?.includes("/docs") && router.asPath.includes("/docs"));
```

Comment:
These routes do not exist in this site – worth a check when copying in code from elsewhere to see if it all remains relevant.